### ISSUE TITLE
Lib Refactoring

### ISSUE THREAD

**Issue by JoeyHwong-gk** [kubeedge/sedna#46]
#### Description

Some of  current components in Sedna Lib  provide services independently based on different features. Each feature provides an independent interface and code structure, which is confusing. To better facilitate those use cases, We need to start thinking about decoupling some of our components, While not a specific action item for Sedna, believe this conversation is worthwhile.

Discussion so far:

1. Secondary development is difficult,  developers cannot replace with their customized processing logic. (e.g. pre-process, post-process, feature engineering, etc.).
2. Too many non-global variables defined in `baseConfig` is that since every function has access to these, it becomes increasingly hard to figure out which feature of sedna actually read and write these variables.
3. Right now Sedna Lib supports only *Tensorflow*, which possibly limits the number of possible users. If we can provide anyway to allowed extend other ML framework, it will benefit more developers.
4. The current feature  modules (`incremental_learning` 、 `joint_inference`、`federated_learning`, etc.) have classes scattered around in a generally unorganized mess，it throws users out of their rhythm when they go to learn 、use different modules. 

#### Details

​	During reviews I've seen a few areas where we can decouple concepts:

1.  By using a  registration of class-factory functions to emulate virtual constructors, developers can invoke different components by change variables in the Config file.

   ```python
   class ClassFactory(object):
       """A Factory Class to manage all class need to register with config."""
   
       __registry__ = {}
   
       @classmethod
       def register(cls, type_name='common', alias=None):
           """Register class into registry.
   
           :param type_name: type_name: type name of class registry
           :param alias: alias of class name
           :return: wrapper
           """
   
           def wrapper(t_cls):
               """Register class with wrapper function.
   
               :param t_cls: class need to register
               :return: wrapper of t_cls
               """
               t_cls_name = alias if alias is not None else t_cls.__name__
               if type_name not in cls.__registry__:
                   cls.__registry__[type_name] = {t_cls_name: t_cls}
               else:
                   if t_cls_name in cls.__registry__:
                       raise ValueError(
                           "Cannot register duplicate class ({})".format(t_cls_name))
                   cls.__registry__[type_name].update({t_cls_name: t_cls})
   
               return t_cls
   
           return wrapper
   
       @classmethod
       def register_cls(cls, t_cls, type_name='common', alias=None):
           """Register class with type name.
   
           :param t_cls: class need to register.
           :param type_name: type name.
           :param alias: class name.
           :return:
           """
           t_cls_name = alias if alias is not None else t_cls.__name__
           if type_name not in cls.__registry__:
               cls.__registry__[type_name] = {t_cls_name: t_cls}
           else:
               if t_cls_name in cls.__registry__:
                   raise ValueError(
                       "Cannot register duplicate class ({})".format(t_cls_name))
               cls.__registry__[type_name].update({t_cls_name: t_cls})
           return t_cls
       
    	@classmethod
       def get_cls(cls, type_name, t_cls_name=None):
           """Get class and bind config to class.
   
           :param type_name: type name of class registry
           :param t_cls_name: class name
           :return:t_cls
           """
           if not cls.is_exists(type_name, t_cls_name):
               raise ValueError("can't find class type {} class name {} in class registry".format(type_name, t_cls_name))
           if t_cls_name is None:
               raise ValueError("can't find class. class type={}".format(type_name))
           t_cls = cls.__registry__.get(type_name).get(t_cls_name)
           return t_cls
   ```

2. Clean up and redesign the base Config class,  each feature  maintains it's specific variables, and ensures that developers can be manually updated the variables.

   ```
   class ConfigSerializable(object):
       """Seriablizable config base class."""
   
       __original__value__ = None
   
       @property
       def __allattr__(self):
           attrs = filter(lambda attr: not (attr.startswith("__") or ismethod(getattr(self, attr))
                                            or isfunction(getattr(self, attr))), dir(self))
           return list(attrs)
   
       def update(self, **kwargs):
           for attr in self.__allattr__:
               if attr not in kwargs:
                   continue
               setattr(self, attr, kwargs[attr])
   
       def to_json(self):
           """Serialize config to a dictionary. It will be very useful in distributed systems. """
           pass
   
       def __getitem__(self, item):
           return getattr(self, item, None)
   
       def get(self, item, default=""):
           return self.__getitem__(item) or default
   
       @classmethod
       def from_json(cls, data):
           """Restore config from a dictionary or a file."""
           pass
   ```

   

3. Refer to tools such as `moxing`、`neuropod` .etc, decouple the ML framework from the features of sedna, allows developers to choose their favorite framework.
   Few frameworks which would be nice to support:

   -  [Scikit learn](https://scikit-learn.org/)
   -  [Pytorch](https://pytorch.org/)
   -  [MindSpore](https://www.mindspore.cn/)

4. Add a common file operation and a unified log-format in `common` module, use an abstract base class to standardize the  feature  modules' interface, and features are invoked by inheriting the base class.

   Specifically, the goals are:

   - Clean up the base module's classes and package structure
   - Create a necessary amount of modules based on big features
   - Revise dependency structure for all the modules

**Comment by ugvddm** [comment]
Hi Joey
we need describe the positioning of sedna and lib firstly.

**Comment by jaypume** [comment]
Hi Joey, it's very nice of you for seeing those problems, but I still have some questions: 
1.  Why does sedna need to support pre-processing and post-processing, while current frameworks are already  integrated with pre-processing function such as `tf.keras.preprocessing`?
2. I want to make sure this is just for sedna's code contributors instead of sedna's AI developer users?
3. It needs to point out what we can abstract from those machine learning frameworks.

**Comment by JoeyHwong-gk** [comment]
> Hi Joey, it's very nice of you for seeing those problems, but I still have some questions:
> 
> 1. Why does sedna need to support pre-processing and post-processing, while current frameworks are already  integrated with pre-processing function such as `tf.keras.preprocessing`?
> 2. I want to make sure this is just for sedna's code contributors instead of sedna's AI developer users?
> 3. It needs to point out what we can abstract from those machine learning frameworks.

Thanks for reply!
1.  Sedna has defined both pre-processing and post-processing in our model training and inference ( clean floder、start rest server 、LC report etc），We would like to be able to do something else between these jobs and model training/inference.
2.  As a deployer, its too waste time to figuring out each variable in global variables, as we're not sure whether a variable is necessary for the feature we're interested in.
3.  Yep! In fact, features of  sedna lib should be framework-agnostic, it should work natively with PyTorch, TensorFlow, and other. For example, model `loading`, `dump_model`, `train`, `inference`, `data_input`, and `output`  etc. should be consistent. Code contributors can develop new features based on the same style.

**Comment by JoeyHwong-gk** [comment]
Rethinking after our Weekly Community Meeting [2021-04-29]，Contributor Experience has been discussing the lib refactoring of sedna and looking into how to make the lib more efficient and maintainable. To summarize the discussions:

1. Sedna formulate policies based on the variables defined in config，some of them should not be maintained by the deployer but by the developer ( hard_example_mind_algorithms、Aggregation_algorithm_in_Federated_learning etc.), We should make it clear to developers **what input parameters they need to maintain**.
2. As our lib is ML-framework-agnostic, developers can define their model (loading, dump_model, train, inference) call methods, and sedna **invokes corresponding methods to implement functions** when required. Of course, sedna can also consider performing operations according to the specified ML framework (train -> keras.model.fit) .
3. I previously thought that the interfaces of each feature should be unified, but because each feature has different objectives, it should be differentiated.
4. As a AI framework of Edge-cloud synergy, lib should open the interfaces for developers access , such as loss functions, optimization functions, and HEM/AGG algorithms etc.

This issue is a spot for further discussion.

**Comment by JoeyHwong-gk** [comment]
> Rethinking after our Weekly Community Meeting [2021-04-29]，Contributor Experience has been discussing the lib refactoring of sedna and looking into how to make the lib more efficient and maintainable. To summarize the discussions:
> 
> 1. Sedna formulate policies based on the variables defined in config，some of them should not be maintained by the deployer but by the developer ( hard_example_mind_algorithms、Aggregation_algorithm_in_Federated_learning etc.), We should make it clear to developers **what input parameters they need to maintain**.
> 2. As our lib is ML-framework-agnostic, developers can define their model (loading, dump_model, train, inference) call methods, and sedna **invokes corresponding methods to implement functions** when required. Of course, sedna can also consider performing operations according to the specified ML framework (train -> keras.model.fit) .
> 3. I previously thought that the interfaces of each feature should be unified, but because each feature has different objectives, it should be differentiated.
> 4. As a AI framework of Edge-cloud synergy, lib should open the interfaces for developers access , such as loss functions, optimization functions, and HEM/AGG algorithms etc.
> 
> This issue is a spot for further discussion.

Examples are as follows:

`### start a api server（for example: rest server in join-inference and ws server in federated-learning）
def run_server():
    job = FederatedLearning()

    server = AggregationServer(
        servername=job.job_name,
        host=job.agg_ip,
        http_port=job.agg_port
    )
    server.start()

def fl_train():
    fl_instance = FederatedLearning(model=CNNModel, # like keras.models
                                    agg="FedAvg"  # define the agg algorithms
                                   )

    # load dataset. data pre-process
    train_data, test_data = TxtDataParse(func=image_process)(
        train_dataset_url, test_dataset_url
    )
    x = np.array([tup[0] for tup in train_data])
    y = np.array([tup[1] for tup in train_data])
	test_x = np.array([tup[0] for tup in test_data])
    
    model = fl_instance.train(x, y, # train ddataset
                      loss='FederatedLoss', # like keras.losses
                      metrics=categorical_accuracy, # keras.metrics
                      optimizer=Adam(learning_rate=0.01), # keras.optimizers
                     )
    pred = model.inference(test_data, post_process=infer_rsl_process)
    return pred



if __name__ == '__main__':
    run_server()
    
    
    `

**Comment by llhuii** [comment]
fixed by #72
/close

feel free to reopen this
